make fails because of "multiple definitions" of util functions

Possible fixes: They should be assigned in a `.cpp` file and not a `.hpp` file. Also there's some other functions that should use the util functions instead of defining their own in order to resolve declarations.

Relevant part of compiler output:
```
/usr/bin/ld: decide/lic1.o: in function `angle_between_points(Coordinate, Coordinate, Coordinate)':
lic1.cpp:(.text+0x0): multiple definition of `angle_between_points(Coordinate, Coordinate, Coordinate)'; decide/lic13.o:lic13.cpp:(.text+0x0): first defined here
/usr/bin/ld: decide/lic1.o: in function `triangleArea(Coordinate, Coordinate, Coordinate)':
lic1.cpp:(.text+0x1a0): multiple definition of `triangleArea(Coordinate, Coordinate, Coordinate)'; decide/lic13.o:lic13.cpp:(.text+0x1a0): first defined here
/usr/bin/ld: decide/lic1.o: in function `contained_in_circle(Coordinate, Coordinate, Coordinate, double)':
lic1.cpp:(.text+0x287): multiple definition of `contained_in_circle(Coordinate, Coordinate, Coordinate, double)'; decide/lic13.o:lic13.cpp:(.text+0x287): first defined here
/usr/bin/ld: decide/lic8.o: in function `angle_between_points(Coordinate, Coordinate, Coordinate)':
lic8.cpp:(.text+0x0): multiple definition of `angle_between_points(Coordinate, Coordinate, Coordinate)'; decide/lic13.o:lic13.cpp:(.text+0x0): first defined here
/usr/bin/ld: decide/lic8.o: in function `triangleArea(Coordinate, Coordinate, Coordinate)':
lic8.cpp:(.text+0x1a0): multiple definition of `triangleArea(Coordinate, Coordinate, Coordinate)'; decide/lic13.o:lic13.cpp:(.text+0x1a0): first defined here
/usr/bin/ld: decide/lic8.o: in function `contained_in_circle(Coordinate, Coordinate, Coordinate, double)':
lic8.cpp:(.text+0x287): multiple definition of `contained_in_circle(Coordinate, Coordinate, Coordinate, double)'; decide/lic13.o:lic13.cpp:(.text+0x287): first defined here
/usr/bin/ld: decide/lic9.o: in function `angle(Coordinate, Coordinate, Coordinate)':
lic9.cpp:(.text+0x0): multiple definition of `angle(Coordinate, Coordinate, Coordinate)'; decide/lic2.o:lic2.cpp:(.text+0x0): first defined here
```

--- a/Makefile
+++ b/Makefile
@@ -26,10 +26,6 @@ format: $(wildcard decide/*.?pp)
 
 
 ### Testing
-# Special case for utils file as it is only a header
-decide/utils.test: decide/utils.test.cpp
-	$(CXX) -o $@ $^ $(CXXFLAGS)
-
 %.test: %.test.cpp %.cpp
 	$(CXX) -o $@ $^ $(CXXFLAGS)
 

--- a/Makefile
+++ b/Makefile
@@ -26,7 +26,7 @@ format: $(wildcard decide/*.?pp)
 
 
 ### Testing
-%.test: %.test.cpp %.cpp
+%.test: %.test.cpp %.cpp decide/utils.cpp
 	$(CXX) -o $@ $^ $(CXXFLAGS)
 
 # Compile all test files together w/ normal file and run them

--- a/decide/lic2.cpp
+++ b/decide/lic2.cpp
@@ -1,4 +1,5 @@
 #include <decide/decide.hpp>
+#include <decide/utils.hpp>
 
 /*There exists at least one set of three consecutive data points which form an
 angle such that: angle < (PI−EPSILON) or angle > (PI+EPSILON) The second of the
@@ -6,17 +7,6 @@ three consecutive points is always the vertex of the angle. If either the first
 point or the last point (or both) coincides with the vertex, the angle is
 undefined and the LIC is not satisfied by those three points. (0 ≤ EPSILON <
 PI)*/
-
-double angle(Coordinate a, Coordinate b, Coordinate c) {
-  double ax = a.x - b.x;
-  double ay = a.y - b.y;
-  double bx = a.x - c.x;
-  double by = a.y - c.y;
-  double top = (ax * bx) + (ay * by);
-  double bottom = sqrt(pow(ax, 2) + pow(ay, 2)) * sqrt(pow(bx, 2) + pow(by, 2));
-  return acos(top / bottom);
-}
-
 bool lic2(Points points, Parameters parameters) {
   if (points.size() < 3 || parameters.EPSILON < 0 || parameters.EPSILON >= pi) {
     return false;
@@ -26,7 +16,7 @@ bool lic2(Points points, Parameters parameters) {
     if (points[i + 1] == points[i] || points[i + 1] == points[i + 2]) {
       continue;
     }
-    double ang = angle(points[i + 1], points[i], points[i + 2]);
+    double ang = angle_between_points(points[i + 1], points[i], points[i + 2]);
     if (ang < (pi - parameters.EPSILON) || ang > (pi + parameters.EPSILON)) {
       return true;
     }

--- a/decide/lic9.cpp
+++ b/decide/lic9.cpp
@@ -1,5 +1,5 @@
 #include <decide/decide.hpp>
-#include <math.h>
+#include <decide/utils.hpp>
 
 /*There exists at least one set of three data points separated by exactly C PTS
 and D PTS consecutive intervening points, respectively, that form an angle such
@@ -8,17 +8,6 @@ of three points is always the vertex of the angle. If either the first point or
 the last point (or both) coincide with the vertex, the angle is undefined and
 the LIC is not satisfied by those three points. When NUMPOINTS < 5, the
 condition is not met. 1 ≤ C PTS, 1 ≤ D PTS C PTS+D PTS ≤ NUMPOINTS−3*/
-double angle(Coordinate a, Coordinate b, Coordinate c) {
-  double ax = a.x - b.x;
-  double ay = a.y - b.y;
-  double bx = a.x - c.x;
-  double by = a.y - c.y;
-  double top = (ax * bx) + (ay * by);
-  double bottom = sqrt(pow(ax, 2) + pow(ay, 2)) * sqrt(pow(bx, 2) + pow(by, 2));
-
-  return acos(top / bottom);
-}
-
 bool lic9(Points points, Parameters parameters) {
   int c = parameters.CPTS + 1;
   int d = parameters.DPTS + 1;
@@ -32,7 +21,8 @@ bool lic9(Points points, Parameters parameters) {
     if (points[i] == points[i + c] || points[i + c] == points[i + c + d])
       continue;
 
-    double ang = angle(points[i + c], points[i], points[i + c + d]);
+    double ang =
+        angle_between_points(points[i + c], points[i], points[i + c + d]);
 
     if (ang < (pi - parameters.EPSILON) || ang > (pi + parameters.EPSILON))
       return true;

--- a/decide/utils.cpp
+++ b/decide/utils.cpp
@@ -1,0 +1,63 @@
+#include <decide/utils.hpp>
+
+double angle_between_points(Coordinate A, Coordinate B, Coordinate C) {
+  if (A.x == B.x && A.y == B.y)
+    return 0.0;
+
+  if (A.x == C.x && A.y == C.y)
+    return 0.0;
+
+  if (B.x == C.x && B.y == C.y)
+    return 0.0;
+
+  double a = atan2(B.y - A.y, B.x - A.x);
+  double b = atan2(C.y - A.y, C.x - A.x);
+  return b - a;
+}
+
+double triangleArea(Coordinate p1, Coordinate p2, Coordinate p3) {
+  // See https://www.omnicalculator.com/math/area-triangle-coordinates
+  return (1.0 / 2.0) * abs(p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) +
+                           p3.x * (p1.y - p2.y));
+}
+
+bool contained_in_circle(Coordinate point1, Coordinate point2,
+                         Coordinate point3, double RADIUS1) {
+  // We first check if the triangle is acute or obtuse
+  double lengthA = point1.distance(point2);
+  double lengthB = point2.distance(point3);
+  double lengthC = point3.distance(point1);
+  if (lengthB > lengthA) {
+    std::swap(lengthB, lengthA);
+  }
+  if (lengthC > lengthA) {
+    std::swap(lengthC, lengthA);
+  }
+
+  // lengthA is now the longest edge, or the hypotenuse
+
+  if (pow(lengthB, 2) + pow(lengthC, 2) < pow(lengthA, 2)) {
+    // The triangle is obtuse and the radius is the longest distance
+    // divided by two (can be shown with sine and cosine of a given angle)
+    double radius = lengthA / 2.0;
+    if (radius <= RADIUS1) {
+      return true;
+    }
+  } else {
+    // The triangle is acute or right. We find the circumcentre using
+    // https://en.wikipedia.org/wiki/Circumscribed_circle#Other_properties
+    // Since this radius marks the equidistant length from all points, and
+    // a non-obtuse triangle has all points on the smallest circle, we
+    // compare this radius to RADIUS1 for the answer
+    double diameter =
+        2 * lengthA * lengthB * lengthC /
+        (sqrt((lengthA + lengthB + lengthC) * (lengthB + lengthC - lengthA) *
+              (lengthA + lengthC - lengthB) * (lengthA + lengthB - lengthC)));
+    double radius = diameter / 2.0;
+    if (radius <= RADIUS1) {
+      return true;
+    }
+  }
+
+  return false;
+}

--- a/decide/utils.hpp
+++ b/decide/utils.hpp
@@ -9,69 +9,14 @@
 
 // Given three points returns the angle formed by the triangle a --> b --> c
 // expressed in radians.
-double angle_between_points(Coordinate A, Coordinate B, Coordinate C) {
-  if (A.x == B.x && A.y == B.y)
-    return 0.0;
-
-  if (A.x == C.x && A.y == C.y)
-    return 0.0;
-
-  if (B.x == C.x && B.y == C.y)
-    return 0.0;
-
-  double a = atan2(B.y - A.y, B.x - A.x);
-  double b = atan2(C.y - A.y, C.x - A.x);
-  return b - a;
-}
+double angle_between_points(Coordinate A, Coordinate B, Coordinate C);
 
 // Calculates the area of a triangle
-double triangleArea(Coordinate p1, Coordinate p2, Coordinate p3) {
-  // See https://www.omnicalculator.com/math/area-triangle-coordinates
-  return (1.0 / 2.0) * abs(p1.x * (p2.y - p3.y) + p2.x * (p3.y - p1.y) +
-                           p3.x * (p1.y - p2.y));
-}
+double triangleArea(Coordinate p1, Coordinate p2, Coordinate p3);
 
 /// Check whether three points can be contained in a circle of
 /// radius RADIUS1
 bool contained_in_circle(Coordinate point1, Coordinate point2,
-                         Coordinate point3, double RADIUS1) {
-  // We first check if the triangle is acute or obtuse
-  double lengthA = point1.distance(point2);
-  double lengthB = point2.distance(point3);
-  double lengthC = point3.distance(point1);
-  if (lengthB > lengthA) {
-    std::swap(lengthB, lengthA);
-  }
-  if (lengthC > lengthA) {
-    std::swap(lengthC, lengthA);
-  }
-
-  // lengthA is now the longest edge, or the hypotenuse
-
-  if (pow(lengthB, 2) + pow(lengthC, 2) < pow(lengthA, 2)) {
-    // The triangle is obtuse and the radius is the longest distance
-    // divided by two (can be shown with sine and cosine of a given angle)
-    double radius = lengthA / 2.0;
-    if (radius <= RADIUS1) {
-      return true;
-    }
-  } else {
-    // The triangle is acute or right. We find the circumcentre using
-    // https://en.wikipedia.org/wiki/Circumscribed_circle#Other_properties
-    // Since this radius marks the equidistant length from all points, and
-    // a non-obtuse triangle has all points on the smallest circle, we
-    // compare this radius to RADIUS1 for the answer
-    double diameter =
-        2 * lengthA * lengthB * lengthC /
-        (sqrt((lengthA + lengthB + lengthC) * (lengthB + lengthC - lengthA) *
-              (lengthA + lengthC - lengthB) * (lengthA + lengthB - lengthC)));
-    double radius = diameter / 2.0;
-    if (radius <= RADIUS1) {
-      return true;
-    }
-  }
-
-  return false;
-}
+                         Coordinate point3, double RADIUS1);
 
 #endif

